infra/src/main/kotlin/fi/fta/geoviite/infra/tracklayout/LayoutAlignmentService.kt

```diff
@@ -32,6 +31,51 @@ class LayoutAlignmentService(
         val alignment = emptyAlignment()
         return alignment to dao.insert(alignment)
     }
+
+    fun getGeometrySectionsByPlan(
+        publishType: PublishType,
+        id: IntId<LayoutAlignment>,
+        trackNumberId: IntId<TrackLayoutTrackNumber>,
+        boundingBox: BoundingBox?
+    ): List<SegmentGeometryAndMetadata> {
+        val segmentGeometriesAndMetadatas = dao.fetchSegmentGeometriesAndPlanMetadata(id)
+
+        val filtered = if (boundingBox != null) filterAndClipSegmentsToBoundingBox(
+            segmentGeometriesAndMetadatas,
+            boundingBox
+        ) else segmentGeometriesAndMetadatas
+        return foldSegmentsByPlan(filtered)
+    }
+}
+
+private fun foldSegmentsByPlan(segments: List<SegmentGeometryAndMetadata>) =
+    segments.fold(mutableListOf<SegmentGeometryAndMetadata>()) { acc, element ->
+        val last = acc.lastOrNull()
+        if (last == null || last.planId != element.planId || last.source != element.source) acc.add(
+            element
+        )
+        else acc.set(acc.lastIndex, last.copy(points = last.points + element.points))
+        acc
+    }
+
+private fun clipSegmentPointsToBoundingBox(segment: SegmentGeometryAndMetadata, boundingBox: BoundingBox) =
+    segment.copy(points = segment.points.filter { boundingBox.contains(it) })
+
+private fun filterAndClipSegmentsToBoundingBox(
+    segments: List<SegmentGeometryAndMetadata>,
+    boundingBox: BoundingBox
+): List<SegmentGeometryAndMetadata> {
+    val firstIndex =
+        segments.indexOfFirst { segment -> segment.points.any{ boundingBox.contains(it) } }
```

Comment:
Ah, no tässä toki käytetään kaikkia pointteja eikä vain päitä. Mutta tätä ei tarvitsisi tehdä, jos hakee bboxilla suoraan kannasta.

Reply:
Sinällään tuolle suodatukselle ja clippailulle on jo logiikka karttaesityksen teossa, tämä menee hiukan duplikaatiksi sen kanssa. Toki data on vähän eri näköinen :thinking: